### PR TITLE
Alternative implementation of pass msve-vector-bits to Flang to embed in .ll files / release_13x

### PR DESCRIPTION
**Pull request by pawosm-arm**
This is a new reimplementation of the PR #84

Corresponding change to the flang-compiler/flang project needs to be applied.

### CODE REVIEW

File: .github/workflows/clang-tests.yml [hunk 1]
```diff
@@ -39,8 +42,16 @@ jobs:
       uses: hendrikmuhs/ccache-action@v1
       with:
         key: ${{ matrix.os }}
-    - name: Test clang
-      uses: llvm/actions/build-test-llvm-project@main
-      with:
-        cmake_args: -G Ninja  -DLLVM_ENABLE_PROJECTS="clang" -DCMAKE_BUILD_TYPE=Release -DCMAKE_C_COMPILER_LAUNCHER=ccache -DCMAKE_CXX_COMPILER_LAUNCHER=ccache
-        build_target: check-clang
+    - name: Test clang macOs
```
Comment (bryanpkc) [inline .github/workflows/clang-tests.yml]: Why isn't the name of the operating system spelled the right way, i.e. `macOS`?

File: clang/test/Driver/flang/emit-flang-attrs.f90 [hunk 2]
```diff
@@ -0,0 +1,57 @@
+! REQUIRES: aarch64-registered-target
```
Comment (bryanpkc) [inline clang/test/Driver/flang/emit-flang-attrs.f90]: Should this test be marked `REQUIRES: classic_flang`? Or will it work with LLVM Flang?

File: clang/test/Driver/flang/flang.f90 [hunk 3]
```diff
@@ -19,32 +19,32 @@
 
 ! Check that f90 files are not treated as "previously preprocessed"
 ! ... in --driver-mode=flang.
-! RUN: %clang --driver-mode=flang -### -E                  %s 2>&1 | FileCheck --check-prefixes=ALL,CHECK-E %s
+! RUN: %flang -### -E                  %s 2>&1 | FileCheck --check-prefixes=ALL,CHECK-E %s
```
Comment (bryanpkc) [inline clang/test/Driver/flang/flang.f90]: This test is marked `UNSUPPORTED: classic_flang`. Did you verify that this will work correctly with LLVM Flang?

**Comment by bryanpkc** [comment]
Can the commits 222b18ff15c1445ef07cc79551e90d83d1225933, 84262af9e398518d0792c9e213ffe5295c762286, 5a9a2229f8274dcc780ecf73a1be0af3ac63a424 be upstreamed in a form that works with both Classic Flang and LLVM Flang?

File: clang/test/Driver/flang/emit-flang-attrs.f90 (hunk 2 above)
Comment (pawosm-arm) [reply clang/test/Driver/flang/emit-flang-attrs.f90]: > Can the commits [222b18f](https://github.com/flang-compiler/classic-flang-llvm-project/commit/222b18ff15c1445ef07cc79551e90d83d1225933), [84262af](https://github.com/flang-compiler/classic-flang-llvm-project/commit/84262af9e398518d0792c9e213ffe5295c762286), [5a9a222](https://github.com/flang-compiler/classic-flang-llvm-project/commit/5a9a2229f8274dcc780ecf73a1be0af3ac63a424) be upstreamed in a form that works with both Classic Flang and LLVM Flang?

No, at least, not by us.

**Comment by pawosm-arm** [comment]
@bryanpkc those tests (with `REQUIRES: aarch64-registered-target`) specifically require AArch64 targets as they refer to AArch64-specific features. We have no plans to push new-flang related things to classic-flang-llvm-project similarly as we we have no plans to publish changes that would ensure such compatibility to already submitted test cases.

File: .github/workflows/clang-tests.yml (hunk 1 above)
Comment (pawosm-arm) [reply .github/workflows/clang-tests.yml]: a typo. fixed.

File: clang/test/Driver/flang/flang.f90 (hunk 3 above)
Comment (pawosm-arm) [reply clang/test/Driver/flang/flang.f90]: No, I don't have any workable LLVM Flang build at my disposal.

**Comment by bryanpkc** [comment]
@pawosm-arm I understand your position. My question is, why is this PR modifying tests that are only exercised for the new Flang? They will not run in a Classic Flang build.

**Comment by pawosm-arm** [comment]
> @pawosm-arm I understand your position. My question is, why is this PR modifying tests that are only exercised for the new Flang? They will not run in a Classic Flang build.

I do understand your point now. Thing is, my part of this PR is just the core logic, the rest I have inherited from other PR's that my code required to be merged to make it fully working. I've reached the author and we agreed that the easiest is to just remove the parts that are not necessary for my code to work. Also, the test that this commit extends should not be moved to the `flang` directory as it is not being executed for the classic flang when it is there.

**Comment by pawosm-arm** [comment]
Hello all,
what is the current reviewer's standpoint on this one?

**Comment by RichBarton-Arm** [comment]
If I am interpreting Bryan's request correctly, I think we need to add `REQUIRES: classic_flang` on all the new testcases added so that they only run when the codebase is built with `ENABLE_CLASSIC_FLANG` is set. They perhaps should also live in the `flang` subdirectory of `clang/test/Driver`. This is a mistake in my original patch that you have inherited here.

Thinking aloud. Perhaps it would be less confusing to put all the classic flang tests into a new subdirectory called `classic_flang`. We could then apply the `REQUIRES`/`UNSUPPORTED` at a top level on each directory rather than in each test individually? One for a future patch though.

File: clang/test/Driver/aarch64-implied-sve-features.c [hunk 4]
```diff
@@ -0,0 +1,78 @@
+// RUN: %clang -target aarch64-linux-gnu -march=armv8-a+sve %s -### 2>&1 | FileCheck %s --check-prefix=SVE-ONLY
```
Comment (bryanpkc) [inline clang/test/Driver/aarch64-implied-sve-features.c:3]: Does this test need `REQUIRES: aarch64-registered-target`?

**Comment by bryanpkc** [comment]
I like @RichBarton-Arm's suggestion.

File: clang/test/Driver/aarch64-implied-sve-features.c (hunk 4 above)
Comment (pawosm-arm) [reply clang/test/Driver/aarch64-implied-sve-features.c:3]: Proposed change makes sense, and I'll add it to this PR, but bear in mind, I'm pulling in here an mainline llvm-project commit.

**Comment by pawosm-arm** [comment]
> I like @RichBarton-Arm's suggestion.

It is a good suggestion, but I think it goes way beyond the scope of this PR. Meanwhile, I've added missing `REQUIRES: classic_flang` directives.

**Review by bryanpkc — APPROVED** [review]
LGTM.

**Pull request merged**